run/opencl/opencl_misc.h

```diff
@@ -393,239 +403,86 @@ inline int check_pkcs_pad(const uchar *data, int len, int blocksize)
 	} while (0)
 
 /*
- * Optimized functions. You need to pick the one that corresponds to the
- * source- and destination memory type(s).
- *
- * Note that for very small sizes, the overhead may make these functions
- * slower than naive code. On the other hand, due to inlining we will
- * hopefully have stuff optimized away more often than not!
+ * memcpy functions.  Until we require OpenCL 2.0, you need to pick the one
+ * that corresponds to the source- and destination memory type(s).
  */
 
 /* src and dst are private mem */
-inline void memcpy_pp(void *dst, const void *src, uint count)
+inline void memcpy_pp(void* restrict dst, const void* restrict src, uint count)
 {
-	union {
-		const uint *w;
-		const uchar *c;
-	} s;
-	union {
-		uint *w;
-		uchar *c;
-	} d;
-
-	s.c = src;
-	d.c = dst;
-
-	if (((size_t)dst & 0x03) == ((size_t)src & 0x03)) {
-		while (((size_t)s.c) & 0x03 && count--)
-			*d.c++ = *s.c++;
-
-		while (count >= 4) {
-			*d.w++ = *s.w++;
-			count -= 4;
-		}
-	}
+	const char *s = src;
+	char *d = dst;
 
-	while (count--) {
-		*d.c++ = *s.c++;
-	}
+	while (count--)
+		*d++ = *s++;
```

Comment:
Is `++` a defined operation for `void *` in OpenCL? (In C, it isn't or at least wasn't - it was a GNU extension.)